linux/src/async_socket_linux.c

```diff
@@ -711,3 +817,78 @@ int async_socket_receive_async(ASYNC_SOCKET_HANDLE async_socket, ASYNC_SOCKET_BU
+{
+    int result;
+
+    // Codes_SRS_ASYNC_SOCKET_LINUX_04_012: [ If async_socket is NULL, async_socket_notify_io_async shall fail and return a non-zero value. ]
+    // Codes_SRS_ASYNC_SOCKET_LINUX_04_013: [ If on_notify_io_complete is NULL, async_socket_notify_io_async shall fail and return a non-zero value. ]
+    // Codes_SRS_ASYNC_SOCKET_LINUX_04_014: [ If io_type has an invalid value, then async_socket_notify_io_async shall fail and return a non-zero value. ]
+    // Codes_SRS_ASYNC_SOCKET_LINUX_04_016: [ on_notify_io_complete_context is allowed to be NULL. ]
+    if (async_socket == NULL || on_notify_io_complete == NULL || (io_type != ASYNC_SOCKET_NOTIFY_IO_TYPE_IN && io_type != ASYNC_SOCKET_NOTIFY_IO_TYPE_OUT))
+    {
+        LogError(
+            "Invalid arguments: ASYNC_SOCKET_HANDLE async_socket=%p, ASYNC_SOCKET_NOTIFY_IO_TYPE io_type=%" PRI_MU_ENUM
+            ", ON_ASYNC_SOCKET_NOTIFY_IO_COMPLETE on_io_complete=%p, void* on_io_complete_context=%p",
+            async_socket, MU_ENUM_VALUE(ASYNC_SOCKET_NOTIFY_IO_TYPE, io_type), on_notify_io_complete, on_notify_io_complete_context);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        // Codes_SRS_ASYNC_SOCKET_LINUX_04_015: [ If the async socket's current state is not ASYNC_SOCKET_LINUX_STATE_OPEN then async_socket_notify_io_async shall fail and return a non-zero value. ]
+        ASYNC_SOCKET_LINUX_STATE current_state;
+        if ((current_state = interlocked_add(&async_socket->state, 0)) != ASYNC_SOCKET_LINUX_STATE_OPEN)
+        {
+            LogWarning("Not open, current state is %" PRI_MU_ENUM "", MU_ENUM_VALUE(ASYNC_SOCKET_LINUX_STATE, current_state));
+            result = MU_FAILURE;
+        }
+        else
+        {
+            (void)interlocked_increment(&async_socket->pending_api_calls);
+
+            // Codes_SRS_ASYNC_SOCKET_LINUX_04_017: [ Otherwise async_socket_notify_io_async shall create a context for the notify where the on_notify_io_complete and on_notify_io_complete_context shall be stored. ]
+            ASYNC_SOCKET_IO_CONTEXT* io_context = malloc(sizeof(ASYNC_SOCKET_IO_CONTEXT));
+            if (io_context == NULL)
+            {
+                // Codes_SRS_ASYNC_SOCKET_LINUX_04_020: [ If any error occurs, async_socket_notify_io_async shall fail and return a non-zero value. ]
+                LogError("failure in malloc(sizeof(ASYNC_SOCKET_IO_CONTEXT)=%zu) failed", sizeof(ASYNC_SOCKET_IO_CONTEXT));
+                result = MU_FAILURE;
+            }
+            else
+            {
+                io_context->io_type = ASYNC_SOCKET_IO_TYPE_NOTIFY;
+                io_context->on_notify_io_complete = on_notify_io_complete;
+                io_context->callback_context = on_notify_io_complete_context;
+                io_context->async_socket = async_socket;
+
+                int epoll_op = (io_type == ASYNC_SOCKET_NOTIFY_IO_TYPE_IN) ? EPOLLIN : EPOLLOUT;
```

Comment:
>int epoll_op = (io_type == ASYNC_SOCKET_NOTIFY_IO_TYPE_IN) ? EPOLLIN : EPOLLOUT; [](http://example.com/codeflow?start=16&length=80)

This may need to change due to our talk today, maybe these will be an or of both values.

Reply:
Added a couple of tasks last week after our discussion (25319277 & 25319272). If this really is a race as we suspect it might be then the fix needs to happen here and, in the send/receive flows.